### PR TITLE
Update Truffle

### PR DESCRIPTION
**Pull request by eregon**
* To Match Sulong's Truffle.

### CODE REVIEW

**Comment by eregon** [comment]
It now looks like cexts are failing with a NPE thrown by Sulong or Truffle Interop.

**Comment by chrisseaton** [comment]
We need to update to the new top-level engine API in Ruby first to fix this.

File: src/main/java/org/truffleruby/options/CommandLineParser.java [hunk 1]
```diff
@@ -156,8 +156,8 @@ private void processArgv() {
             }
         }
         // Remaining arguments are for the script itself
-        arglist.addAll(Arrays.asList(OptionsCatalog.ARGUMENTS.checkValue(config.getOptions().get(OptionsCatalog.ARGUMENTS.getName()))));
-        config.getOptions().put(OptionsCatalog.ARGUMENTS.getName(), arglist.toArray(new String[arglist.size()]));
+        //arglist.addAll(Arrays.asList(OptionsCatalog.ARGUMENTS.checkValue(config.getOptions().get(OptionsCatalog.ARGUMENTS.getName()))));
```
Comment (eregon) [inline src/main/java/org/truffleruby/options/CommandLineParser.java]: I'd rather just remove the code if it's not working anymore.
This used to allow setting args as a Java property and concatenate them with CLI args.
I don't know of any use-case.

File: src/main/java/org/truffleruby/options/CommandLineParser.java [hunk 2]
```diff
@@ -335,7 +335,7 @@ private void processArgument() {
                     }
                 case 'x':
                     disallowedInRubyOpts(argument);
-                    config.getOptions().put(OptionsCatalog.IGNORE_LINES_BEFORE_RUBY_SHEBANG.getName(), true);
+                    config.getOptions().put(OptionsCatalog.IGNORE_LINES_BEFORE_RUBY_SHEBANG.getName(), Boolean.TRUE.toString());
```
Comment (eregon) [inline src/main/java/org/truffleruby/options/CommandLineParser.java]: This is really ugly, I would rather go for just the literal `"true"`.

File: src/main/java/org/truffleruby/options/OptionsBuilder.java [hunk 3]
```diff
@@ -48,15 +49,15 @@ public void set(Map<String, Object> properties) {
 
     public void set(OptionValues optionValues) {
         for (OptionDescription<?> option : OptionsCatalog.allDescriptions()) {
-            // TODO CS 19-Jun-17 default value of null isn't supported
-
-            if (option.getDefaultValue() != null) {
-                set(option.getName(), optionValues.get(optionValues.getDescriptors().get(option.getSDKName()).getKey()));
-            }
+            set(option.getName(), optionValues.get(optionValues.getDescriptors().get(option.getSDKName()).getKey()));
```
Comment (eregon) [inline src/main/java/org/truffleruby/options/OptionsBuilder.java]: Please extract the key to a local variable.

File: src/main/java/org/truffleruby/options/OptionsBuilder.java [hunk 4]
```diff
@@ -48,15 +49,15 @@ public void set(Map<String, Object> properties) {
...
         }
     }
 
     private void set(String name, Object value) {
+        if (name.startsWith(RubyLanguage.ID + ".")) {
+            name = name.substring(RubyLanguage.ID.length() + 1, name.length());
```
Comment (eregon) [inline src/main/java/org/truffleruby/options/OptionsBuilder.java]: String#substring does not need a second argument.

File: src/main/java/org/truffleruby/core/CoreLibrary.java [hunk 5]
```diff
@@ -1030,7 +1023,8 @@ public void initializeEncodingManager() {
             getContext().getEncodingManager().setDefaultExternalEncoding(getContext().getEncodingManager().getLocaleEncoding());
         }
 
-        final String internalEncodingName = getContext().getOptions().INTERNAL_ENCODING;
+        String internalEncodingName = getContext().getOptions().INTERNAL_ENCODING;
+        internalEncodingName = internalEncodingName.isEmpty() ? null : internalEncodingName;
         if (internalEncodingName != null && !internalEncodingName.isEmpty()) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: The 2 lines above can all be simplified to `if !internalEncodingName.isEmpty()`.

File: spec/tags/command_line/dash_x_tags.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+fails:The -x command line option fails when /#!.*ruby.*/-ish line in target file is not found
```
Comment (eregon) [inline spec/tags/command_line/dash_x_tags.txt]: What is this new failure?

File: src/main/java/org/truffleruby/core/CoreLibrary.java (hunk 5 above)
Comment (eregon) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: same for externalEncoding I suspect.

File: src/main/java/org/truffleruby/Main.java [hunk 7]
```diff
@@ -97,14 +100,15 @@ public static void main(String[] args) {
 
         if (config.getShouldRunInterpreter()) {
             final String filename = displayedFileName(config);
-            final PolyglotEngine engine = createPolyglotEngine(config, filename);
-            final RubyContext context = loadContext(engine);
+            final Engine engine = createEngine(config, filename);
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/Main.java]: Done.

File: src/main/java/org/truffleruby/Main.java [hunk 8]
```diff
@@ -274,14 +273,11 @@ private static boolean checkFileSyntax(PolyglotEngine engine, RubyContext contex
...
-    private static boolean runCheckSyntax(PolyglotEngine engine, RubyContext context, InputStream in, String filename) {
+    private static boolean runCheckSyntax(PolyglotContext polyglotContext, RubyContext context, InputStream in, String filename) {
         context.setSyntaxCheckInputStream(in);
         context.setOriginalInputFile(filename);
 
-        return engine.eval(loadBootSource(
-                //language=ruby
-                "Truffle::Boot.check_syntax",
-                "check_syntax")).as(Boolean.class);
+        return polyglotContext.eval(RubyLanguage.ID, Source.newBuilder("Truffle::Boot.check_syntax").name("check_syntax").build()).asBoolean();
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/Main.java]: Fixed.

File: src/main/java/org/truffleruby/language/CallStackManager.java [hunk 9]
```diff
@@ -270,7 +271,7 @@ public Object visitFrame(FrameInstance frameInstance) {
 
         if (!activations.isEmpty()) {
             final Activation last = activations.get(activations.size() - 1);
-            if (last.getCallNode().getRootNode().getSourceSection().getSource() == context.getCoreLibrary().getMainBootSource()) {
+            if (last.getCallNode().getRootNode().getSourceSection().getSource().getName().equals(CoreLibrary.MAIN_BOOT_SOURCE_NAME)) {
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/language/CallStackManager.java]: Done.

File: src/main/java/org/truffleruby/options/CommandLineParser.java (hunk 1 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/options/CommandLineParser.java]: Done.

File: src/main/java/org/truffleruby/options/CommandLineParser.java (hunk 2 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/options/CommandLineParser.java]: Done.

File: src/main/java/org/truffleruby/options/OptionsBuilder.java (hunk 3 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/options/OptionsBuilder.java]: Most of this code will go in the next commit when I properly switch to the new options.

File: src/main/java/org/truffleruby/options/OptionsBuilder.java (hunk 4 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/options/OptionsBuilder.java]: Done.

File: src/main/java/org/truffleruby/core/CoreLibrary.java (hunk 5 above)
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/CoreLibrary.java]: Done.

File: spec/tags/command_line/dash_x_tags.txt (hunk 6 above)
Comment (chrisseaton) [inline spec/tags/command_line/dash_x_tags.txt]: Fixed.

File: src/main/java/org/truffleruby/Main.java [hunk 10]
```diff
@@ -100,8 +100,8 @@ public static void main(String[] args) {
 
         if (config.getShouldRunInterpreter()) {
             final String filename = displayedFileName(config);
-            try (final Engine engine = createEngine(config, filename);
-                 final PolyglotContext polyglotContext = engine.newPolyglotContextBuilder().setArguments(RubyLanguage.ID, config.getArguments()).build();) {
+            try (Engine engine = createEngine(config, filename);
+                 PolyglotContext polyglotContext = engine.newPolyglotContextBuilder().setArguments(RubyLanguage.ID, config.getArguments()).build()) {
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/Main.java]: Why did you make these mutable?

Comment (eregon) [inline src/main/java/org/truffleruby/Main.java]: It's less noise visually, and the `try` already shows we are declaring a new variable.

**Pull request merged**